Refine migration support for durable objects

xortive opened the pull request:
This PR adds support for rename and transfer migrations

xortive commented on src/util.rs:
```diff
@@ -0,0 +1,10 @@
+}
+
+impl<'a, 'b> ApplyToApp<'a, 'b> for App<'a, 'b> {
+    fn apply(self, f: fn(App<'a, 'b>) -> App<'a, 'b>) -> App<'a, 'b> {
```
the crab language is hard sometimes

caass replied:
i love the smell of lifetimes in the morning...

xortive replied:
This apply thing would be pretty useful for cleaning up main. I added it since I wanted to re-use the setup for the command line arguments in my test

caass replied:
yeah i left a comment, def think this should become a trait in a future pr

xortive commented:
@a-robinson particularly interested in your review of the migration coalescing algorithm
@ags799 @caass tagged y'all for the more wrangler-specific pieces of this PR.

ags799 commented on src/settings/toml/migrations.rs:
```diff
@@ -0,0 +1,499 @@
+                let scripts = {
+                    let res: ListScriptsV4ApiResponse = res.json()?;
+                    res.result
+                };
```
purely up to you, but I think starting a new scope here is kind of weird, expected to see
```
let res = ...
let res_json: ... = res.json()?;
let scripts = res.result;
```
but it really doesn't matter

xortive replied:
I quite like using scopes to avoid having to name intermediate results

nilslice replied:
correct me if I'm mistaken or don't have the full context, but can't you avoid the intermediate result and still try the result like this? 

```rust
let scripts = res.json::<ListScriptsV4ApiResponse>()?.result;
```

(not tested !)

ags799 commented on src/settings/toml/migrations.rs:
```diff
@@ -0,0 +1,499 @@
+                    Some(ApiScript {
+                        name: _,
+                        migration_tag: None,
+                    }) => Self::coalesce_migrations(migrations, None),
```
you could use `|` to merge these cases

xortive replied:
good catch, will do

caass commented on src/commands/dev/mod.rs:
```diff
@@ -29,8 +29,6 @@ pub fn dev(
     // before serving requests we must first build the Worker
     build_target(&target)?;
 
-    deploy::pre_upload()?;
-
```
Nice

xortive replied:
I'm still iffy on whether or not it would have been nicer to keep this, or to do all the hackery I did to get a GlobalUser to pass into `api_migrations()`.

caass commented on src/settings/toml/manifest.rs:
```diff
@@ -336,8 +340,12 @@ impl Manifest {
             durable_objects: self.durable_objects.clone(), // Not inherited
-            site: self.site.clone(), // Inherited
-            vars: self.vars.clone(), // Not inherited
+            migrations: self // Top level
+                .migrations
+                .clone()
+                .map(|migrations| Migrations { migrations }),
```
Wait...is this gonna work if migrations is `None`? I assume it will but I don't see any like, filtering

xortive replied:
yep, map will pass `None` through

caass commented on src/settings/toml/migrations.rs:
```diff
@@ -0,0 +1,499 @@
+            Arg::with_name("new_class")
+            .help("allow durable objects to be created from a class in your script")
+            .long("new-class")
+            .takes_value(true)
+            .number_of_values(1)
+            .value_name("class name")
+            .multiple(true)
+        )
+        .arg(
+            Arg::with_name("delete_class")
+            .help("delete all durable objects associated with a class in your script")
+            .long("delete-class")
+            .takes_value(true)
+            .number_of_values(1)
+            .value_name("class name")
+            .multiple(true)
+        )
+        .arg(
+            Arg::with_name("rename_class")
+            .help("rename a durable object class")
+            .long("rename-class")
+            .takes_value(true)
+            .number_of_values(2)
+            .value_names(&["from class", "to class"])
+            .multiple(true)
+        )
+        .arg(
+            Arg::with_name("transfer_class")
+            .help("transfer all durable objects associated with a class in another script to a class in this script")
+            .long("transfer-class")
+            .takes_value(true)
+            .number_of_values(3)
+            .value_names(&["from script", "from class", "to class"])
+            .multiple(true)
+        )
+        .group(
+            ArgGroup::with_name("migrations")
+            .args(&["new_class", "delete_class", "rename_class", "transfer_class"])
+            .multiple(true)
+        )
```
dude this pattern is so good...we should make this a trait and impl for subcommands in a future pr

caass reviewed (commented):
I'd like to understand what this changes in terms of wrangler usage. This basically adds a [migrations] tag to wrangler.toml? and then you update that every time you migrate?

xortive commented:
> I'd like to understand what this changes in terms of wrangler usage. This basically adds a [migrations] tag to wrangler.toml? and then you update that every time you migrate?

yup! it's a top-level array of migrations, every time you add/rename/transfer classes, you'd add a new entry to the [[migrations]] array

nilslice commented:
sorry, laptop keyboard went crazy and issued some clicks.. didn't mean to add the `blocked` label 😨

xortive commented:
fixed some logic bugs with the input to migration coalescing I discovered after testing:

- duplicate migration tags are properly handled
- removed special case for a single migration that incorrectly attempted to push it up with no tags
- the old migration in `wrangler.toml` was incorrectly being included as part of coalescing. we now start at the next migration

a-robinson commented on src/settings/toml/migrations.rs:
```diff
@@ -0,0 +1,504 @@
+
+        for mc in &migrations[1..] {
+            // Renames will modify existing new, rename, and transfer migrations if they are present
+            for rename in mc.migration.durable_objects.renamed_classes.iter() {
```
If someone has multiple rename migrations away from the same class name (which is also a target of one of the prior migrations), then this will probably lead to weird error messages from the API (along the lines of "class doesn't exist"). But that should be quite uncommon, and it wouldn't be easy to do better here without a bunch more error checking that's already done in the API.

So this looks good 👍

a-robinson replied:
Although I suppose some of this logic could go awry if the initial set of migrations is invalid. For example, if there are multiple matches for a given classname, won't this code only find and update the first? It's not valid to have multiple migrations targeting the same class, but unless wrangler has already validated that then it isn't really safe to assume here.

I think we either need to do that validation or attempt to update all matching migrations in this logic rather than just the first one. I'd personally suggest doing the former rather than the latter, since it seems simpler and more self-contained. You could probably just create a `set`, and as you process each individual migration (including the first batch of them above this loop), make sure its destination class isn't already the destination of a different migration.

a-robinson commented on src/settings/toml/migrations.rs:
```diff
@@ -0,0 +1,504 @@
+                    .iter_mut()
+                    .position(|existing| &existing.to == delete)
+                {
+                    renamed_classes.remove(i);
```
We need to apply the delete migration to the `from` class at this point, right? Otherwise it may stick around when it's supposed to be deleted.

xortive replied:
oo, good point

a-robinson commented on src/settings/toml/migrations.rs:
```diff
@@ -0,0 +1,504 @@
+                    .iter_mut()
+                    .position(|existing| &existing.to == delete)
+                {
+                    transferred_classes.remove(i);
```
I think this is just straight-up invalid and should error out. Removing the transfer migration is definitely not the same as transferring a class and then deleting it, but there's also no way to convert this into a logically equivalent migration that deletes the class from the other script.

a-robinson commented on src/settings/toml/migrations.rs:
```diff
@@ -0,0 +1,504 @@
+    }
+
+    #[test]
+    fn migration_coalescing() {
```
I think we'd benefit from some tests that have invalid migrations in them.

a-robinson commented on src/settings/toml/migrations.rs:
```diff
@@ -0,0 +1,504 @@
+            // deletes will remove existing new, rename, and transfer migrations if they are present
+            for delete in mc.migration.durable_objects.deleted_classes.iter() {
+                if let Some(i) = new_classes.iter().position(|c| c == delete) {
+                    new_classes.remove(i);
```
Hm, the more I read and comment on this logic the more worried I am that it's going to be very hard to get right here... What if there already is a class in the script (as previously stored in the API) that has durable objects on it? This will behave very differently than if we actually sent all the migrations to the server -- if we sent one batch at a time, it would be an error to pass a new-class migration for a class that already exists. But if we aggregate them like this, then the upload will silently succeed with the old class still in place.

I'm curious if you have ideas, but I see two options to avoid this sort of problem:

1. When we get the migration tag for the script, also get the list of DO classes currently implemented by the script and then use them to avoid problems like this. We'll have to add this to the API, but that's fine -- it'll be useful for other purposes anyway. What might be a bit less fun is duplicating all of the validation logic here, but it wouldn't be terribly hard.
2. Change the API to support passing an array of migration batches. I'm not sure this would actually reduce the amount of new validation code we need to write, since handling multiple batches at once will require a bunch of new logic will be required in the API -- what's there today isn't really structured for that.

So I think I'd lean towards the first option.

xortive replied:
Yeah, I'm starting to agree with you as well here. I dislike the idea of re-implementing the validation logic, is there a particular reason you favor 1. over 2.? If it's going to be a similar amount of work, 2. would have the benefit of non-wrangler users getting to upload multiple migrations as well

a-robinson replied:
I suppose the main reasons I had in mind are

1. We'll either have to break the existing API (and thus anyone using the current wrangler RC) in order to accept batches of migrations) or put in a hacky compatibility layer in the parsing of the request to deal with some clients sending batches and others assuming that only one batch (not in an array) can be sent, but given that I suppose such a layer is possible it isn't that big of a deal.
2. I don't think it's very likely that anyone other than wrangler will find the option to send multiple rounds of migrations useful, and we'd be making our API a little tougher to use for anyone who only ever sends a single set of migrations at a time.  I don't get the impression we're likely to have many people programming directly against the API, though.

Neither of those points is particularly strong, so I'm left not really caring all that much either way. We can put it in the API if that's your preference.

xortive commented:
since migration coalescing / migration in toml seems to be somewhat issue-prone as currently implemented, I'll break out a separate PR for just rename/transfer migrations

xortive commented:
Added the aliases for durable objects classes section. in next RC we will make this a breaking change and remove support for the old name

before:
```toml
[durable_objects]
classes = [{ binding = "Counter", class_name = "Counter" }]
```

```toml
[durable_objects]
# binding = "Counter" would be accepted too, but it feels weird to duplicate "binding" in this manner. We should use name instead IMO, but I don't feel strongly if we want to keep using "binding"
bindings = [{ name = "Counter", class_name = "Counter" }]
```

a-robinson reviewed (approved):
LGTM assuming you confirmed the renamed fields work.
